Generic worker

severo opened the pull request:
This PR allows a worker to process different job types. We can still dedicate workers to a sublist of jobs using a comma-separated list of the jobs in `WORKER_LOOP_ONLY_JOB_TYPES`.

This way, we will be able to reduce the allocated but unused resources.

As you can see in chart/env/prod.yaml, I let all the previous workers but added 20 instances of "genericWorker" that can process any job. If it works well, we can reduce the number of specific workers (maybe just /first-rows and /parquet-and-dataset-info) and increase the number of generic ones.

Fixes #737.

HuggingFaceDocBuilder commented:
The docs for this PR live [here](https://moon-ci-docs.huggingface.co/docs/datasets-server/pr_802). All of your documentation changes will be reflected on that endpoint.

AndreaFrancis commented on libs/libcommon/src/libcommon/queue.py:
```diff
@@ -430,25 +497,26 @@ def is_job_in_process(self, dataset: str, config: Optional[str] = None, split: O
 
-    def cancel_started_jobs(self) -> None:
+    def cancel_started_jobs(self, job_type: str) -> None:
         """Cancel all started jobs."""
```
```suggestion
        """Cancel all started jobs for a given type."""
```

AndreaFrancis commented on libs/libcommon/tests/test_queue.py:
```diff
@@ -31,30 +33,30 @@ def test__add_job() -> None:
-    queue._add_job(dataset=test_dataset)
-    assert queue.is_job_in_process(dataset=test_dataset) is True
+    queue._add_job(job_type=test_type, dataset=test_dataset)
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is True
```
```suggestion
    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset)
```

AndreaFrancis commented on libs/libcommon/tests/test_queue.py:
```diff
@@ -31,30 +33,30 @@ def test__add_job() -> None:
     assert job_info["split"] is None
     assert job_info["force"] is True
-    assert queue.is_job_in_process(dataset=test_dataset) is True
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is True
```
```suggestion
    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset)
```

AndreaFrancis commented on libs/libcommon/tests/test_queue.py:
```diff
@@ -31,30 +33,30 @@ def test__add_job() -> None:
     queue.finish_job(job_id=job_info["job_id"], finished_status=Status.SUCCESS)
     # the queue is not empty
-    assert queue.is_job_in_process(dataset=test_dataset) is True
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is True
```
```suggestion
    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset)
```

AndreaFrancis commented on libs/libcommon/tests/test_queue.py:
```diff
@@ -68,7 +70,7 @@ def test__add_job() -> None:
     queue.finish_job(job_id=job_info["job_id"], finished_status=Status.SUCCESS)
     # the queue is empty
-    assert queue.is_job_in_process(dataset=test_dataset) is False
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is False
```
```suggestion
    assert not queue.is_job_in_process(job_type=test_type, dataset=test_dataset)
```

AndreaFrancis commented on libs/libcommon/tests/test_queue.py:
```diff
@@ -78,35 +80,35 @@ def test_upsert_job() -> None:
-    queue.upsert_job(dataset=test_dataset)
-    assert queue.is_job_in_process(dataset=test_dataset) is True
+    queue.upsert_job(job_type=test_type, dataset=test_dataset)
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is True
```
```suggestion
    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset)
```

AndreaFrancis commented on libs/libcommon/tests/test_queue.py:
```diff
@@ -78,35 +80,35 @@ def test_upsert_job() -> None:
     queue.finish_job(job_id=job_info["job_id"], finished_status=Status.SUCCESS)
     # the queue is empty
-    assert queue.is_job_in_process(dataset=test_dataset) is False
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is False
```
```suggestion
    assert not queue.is_job_in_process(job_type=test_type, dataset=test_dataset)
```

AndreaFrancis commented on libs/libcommon/tests/test_queue.py:
```diff
@@ -152,16 +156,20 @@ def test_max_jobs_per_namespace(max_jobs_per_namespace: Optional[int]) -> None:
-    queue.upsert_job(dataset=test_dataset, config=test_config, split="split3")
+    queue = Queue(max_jobs_per_namespace=max_jobs_per_namespace)
+    queue.upsert_job(job_type=test_type, dataset=test_dataset, config=test_config, split="split1")
+    assert (
```
```suggestion
    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset, config=test_config, split="split1")
```

AndreaFrancis commented on libs/libcommon/tests/test_queue.py:
```diff
@@ -152,16 +156,20 @@ def test_max_jobs_per_namespace(max_jobs_per_namespace: Optional[int]) -> None:
     assert job_info["split"] == "split1"
-    assert queue.is_job_in_process(dataset=test_dataset, config=test_config, split="split1") is True
+    assert (
+        queue.is_job_in_process(job_type=test_type, dataset=test_dataset, config=test_config, split="split1") is True
```
```suggestion
        queue.is_job_in_process(job_type=test_type, dataset=test_dataset, config=test_config, split="split1")
```

AndreaFrancis commented on libs/libcommon/tests/test_queue.py:
```diff
@@ -175,43 +183,67 @@ def test_max_jobs_per_namespace(max_jobs_per_namespace: Optional[int]) -> None:
+    test_dataset = "test_dataset"
+    queue = Queue(max_jobs_per_namespace=100)
+    queue.upsert_job(job_type=job_type, dataset=test_dataset, config=None, split=None)
+    assert queue.is_job_in_process(job_type=job_type, dataset=test_dataset, config=None, split=None) is True
```
```suggestion
    assert queue.is_job_in_process(job_type=job_type, dataset=test_dataset, config=None, split=None)
```

AndreaFrancis commented on libs/libcommon/tests/test_queue.py:
```diff
@@ -175,43 +183,67 @@ def test_max_jobs_per_namespace(max_jobs_per_namespace: Optional[int]) -> None:
+    queue = Queue(max_jobs_per_namespace=100)
+    queue.upsert_job(job_type=job_type, dataset=test_dataset, config=None, split=None)
+    assert queue.is_job_in_process(job_type=job_type, dataset=test_dataset, config=None, split=None) is True
+    if only_job_types is not None and job_type not in only_job_types:
```
```suggestion
    if only_job_types and job_type not in only_job_types:
```

AndreaFrancis commented on libs/libcommon/tests/test_queue.py:
```diff
@@ -175,43 +183,67 @@ def test_max_jobs_per_namespace(max_jobs_per_namespace: Optional[int]) -> None:
+    assert queue.is_job_in_process(job_type=job_type, dataset=test_dataset, config=None, split=None) is True
+    if only_job_types is not None and job_type not in only_job_types:
+        with pytest.raises(EmptyQueueError):
+            job_info = queue.start_job(only_job_types=only_job_types)
```
Do you use job_info?

severo replied:
removing, thanks

AndreaFrancis commented on libs/libcommon/tests/test_queue.py:
```diff
@@ -78,35 +80,35 @@ def test_upsert_job() -> None:
     assert job_info["split"] is None
     assert job_info["force"] is True  # the new job inherits from waiting forced jobs
-    assert queue.is_job_in_process(dataset=test_dataset) is True
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is True
```
```suggestion
    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset)
```

AndreaFrancis commented on libs/libcommon/tests/test_queue.py:
```diff
@@ -78,35 +80,35 @@ def test_upsert_job() -> None:
     queue.finish_job(job_id=job_info["job_id"], finished_status=Status.SUCCESS)
     # the queue is not empty
-    assert queue.is_job_in_process(dataset=test_dataset) is True
+    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset) is True
```
```suggestion
    assert queue.is_job_in_process(job_type=test_type, dataset=test_dataset)
```

AndreaFrancis commented on services/worker/tests/test_loop.py:
```diff
@@ -52,18 +51,21 @@ def test_process_next_job(
-    loop.queue.upsert_job(dataset=dataset, config=config, split=split)
-    loop.queue.is_job_in_process(dataset=dataset, config=config, split=split) is True
+    loop.queue.upsert_job(job_type=test_processing_step.job_type, dataset=dataset, config=config, split=split)
+    loop.queue.is_job_in_process(
```
```suggestion
    assert loop.queue.is_job_in_process(
```

AndreaFrancis commented on services/worker/tests/test_loop.py:
```diff
@@ -52,18 +51,21 @@ def test_process_next_job(
+    ) is True
     assert loop.process_next_job() is True
-    loop.queue.is_job_in_process(dataset=dataset, config=config, split=split) is False
+    loop.queue.is_job_in_process(
```
```suggestion
    assert loop.queue.is_job_in_process(
```

severo commented:
Thanks! I fixed based on your comments (and even removed all occurrences of `is False` or `is True`). I rebased on main after https://github.com/huggingface/datasets-server/pull/805, so there should be new changes.
